qa/L0_cuda_shared_memory/cuda_shared_memory_test.py

```diff
@@ -0,0 +1,141 @@
+from tensorrtserver.api import *
+import numpy as np
+import threading
+
+# Raises error since invalid cuda shm region
+try:
+    shm_op0_handle = cshm.create_shared_memory_region("dummy_data", -1, 0)
+except Exception as ex:
+    assert str(ex) == "unable to create cuda shared memory handle"
+
+shared_memory_ctx = SharedMemoryControlContext("localhost:8000",  ProtocolType.HTTP, verbose=False)
+
+# Create a valid cuda shared memory region
+shm_op0_handle = cshm.create_shared_memory_region("dummy_data", 8, 0)
+# Fill data in cuda shared memory region
+cshm.set_shared_memory_region(shm_op0_handle, [np.array([1,2], dtype=np.float32)])
+# Unregister before register does not fail - does nothing
+shared_memory_ctx.unregister(shm_op0_handle)
+# Test if register is working
+shared_memory_ctx.cuda_register(shm_op0_handle)
+
+# Raises error if registering already registered region
+try:
+    shared_memory_ctx.cuda_register(shm_op0_handle)
+except Exception as ex:
+    assert "shared memory block 'dummy_data' already in manager" in str(ex)
+
+# unregister after register
+shared_memory_ctx.unregister(shm_op0_handle)
+cshm.destroy_shared_memory_region(shm_op0_handle)
+
+shm_op0_handle = cshm.create_shared_memory_region("output0_data", 64, 0)
+shm_op1_handle = cshm.create_shared_memory_region("output1_data", 64, 0)
+shm_ip0_handle = cshm.create_shared_memory_region("input0_data", 64, 0)
+shm_ip1_handle = cshm.create_shared_memory_region("input1_data", 64, 0)
+input0_data = np.arange(start=0, stop=16, dtype=np.int32)
+input1_data = np.ones(shape=16, dtype=np.int32)
+cshm.set_shared_memory_region(shm_ip0_handle, [input0_data])
+cshm.set_shared_memory_region(shm_ip1_handle, [input1_data])
+shared_memory_ctx.cuda_register(shm_ip0_handle)
+shared_memory_ctx.cuda_register(shm_ip1_handle)
+shared_memory_ctx.cuda_register(shm_op0_handle)
+shared_memory_ctx.cuda_register(shm_op1_handle)
+
+infer_ctx = InferContext("localhost:8000", ProtocolType.HTTP, "simple", -1, verbose=False)
+def basic_inference(shm_ip0_handle, shm_ip1_handle, shm_op0_handle, shm_op1_handle, error_msg):
+    try:
+        results = infer_ctx.run({ 'INPUT0' : shm_ip0_handle, 'INPUT1' : shm_ip1_handle, },
+                { 'OUTPUT0' : (InferContext.ResultFormat.RAW, shm_op0_handle),
+                'OUTPUT1' : (InferContext.ResultFormat.RAW, shm_op1_handle)}, 1)
+        assert (results['OUTPUT0'][0] == (input0_data + input1_data)).all()
+    except Exception as ex:
+        error_msg.append(str(ex.message()))
+
+# Unregister during inference - inference fails and unregisters
+error_msg = []
+threads = []
+threads.append(threading.Thread(target=basic_inference,
+    args=(shm_ip0_handle, shm_ip1_handle, shm_op0_handle, shm_op1_handle, error_msg)))
+threads.append(threading.Thread(target=shared_memory_ctx.unregister, args=(shm_op0_handle,)))
+for t in threads:
+    t.start()
+for t in threads:
+    t.join()
+
+assert error_msg[0] == "shared memory block 'output0_data' not found in manager"
```

Comment:
Be careful if you want to do parallel work here. Your assetion here actually assume that the unregister happens before the inference is running, the error message is returned when GRPC frontend is trying to get the shared memory block. What if unregister happens after the inference is initiated? The server gets the underlying address of the shared memory block and tries to read from / write to the address, but the address is invalid as it is unmapped. What will the server do in this case?

Reply:
We recommend the user to not unregister a region that is currently being used for inference. The same applies to system shared memory. I will revisit the same in the future. Filing a different ticket for the this. These tests aim to test basic functionality of the python client API